build-tools/packages/build-cli/src/commands/generate/typetests.ts

```diff
@@ -0,0 +1,377 @@
+ * Generated by fluid-type-test-generator in @fluidframework/build-tools.
+ */
+
+import type * as old from "${previousPackageName}${
+				previousPackageLevel === ApiLevel.public ? "" : `/${previousPackageLevel}`
+			}";
+import type * as current from "../../index.js";
+		`.trim(),
+			typeOnly,
+		];
+
+		const testCases = generateCompatibilityTestCases(
+			previousData,
+			currentTypeMap,
+			currentPackageJson,
+			fileHeader,
+		);
+
+		mkdirSync(outDir, { recursive: true });
+
+		writeFileSync(typeTestOutputFile, testCases.join("\n"));
+		this.info(`Generated type test file: ${path.resolve(typeTestOutputFile)}`);
+	}
+}
+
+/**
+ * Tries to find the path to types for a given API level, falling back to another API level (typically public) if the
+ * requested one is not found. The paths returned are relative to the package.
+ */
+function getTypesPathWithFallback(
+	packageJson: PackageJson,
+	level: ApiLevel,
+	fallbackLevel?: ApiLevel,
+): { typesPath: string; levelUsed: ApiLevel } {
+	let chosenLevel: ApiLevel = level;
+	// First try the requested paths, but fall back to public otherwise if configured.
+	let typesPath: string | undefined = getTypesPathFromPackage(packageJson, level);
+
+	if (typesPath === undefined) {
+		// Try the public types if configured to do so. If public types are found adjust the level accordingly.
+		typesPath =
+			fallbackLevel === undefined
+				? undefined
+				: getTypesPathFromPackage(packageJson, fallbackLevel);
+		chosenLevel = fallbackLevel ?? level;
+		if (typesPath === undefined) {
+			throw new Error(
+				`No type definitions found for "${chosenLevel}" API level in ${packageJson.name}`,
+			);
+		}
+	}
+	return { typesPath: typesPath, levelUsed: chosenLevel };
+}
+
+/**
+ * Finds the path to the types of a package using the package's export map or types/typings field.
+ * If the path is found, it is returned. Otherwise it returns undefined.
+ *
+ * This implementation uses resolve.exports to resolve the path to types for a level.
+ *
+ * @param packageJson - The package.json object to check for types paths.
+ * @param level - An API level to get types paths for.
+ * @returns A package relative path to the types.
+ */
+export function getTypesPathFromPackage(
+	packageJson: PackageJson,
+	level: ApiLevel,
+): string | undefined {
+	const entrypoint = level === ApiLevel.public ? "." : `./${level}`;
+
+	// resolve.exports sets some conditions by default, so the ones we supply supplement the defaults. For clarity the
+	// applied conditions are noted in comments.
+	let typesPath: string | undefined;
+	try {
+		// First try to resolve with the "import" condition, assuming the package is either ESM-only or dual-format.
+		// conditions: ["default", "types", "import", "node"]
+		const exports = resolve.exports(packageJson, entrypoint, { conditions: ["types"] });
+		typesPath =
+			exports === undefined || exports.length === 0
+				? packageJson.types ?? packageJson.typings
+				: exports[0];
```

Comment:
This is a mistake. The fallback to types/typings was only intended to be done once, after trying to resolve with exports using import and require conditions. The fallback might not be needed for more recent versions of packages where we have exports defined, but it seems worth keeping for now just in case. I'll refactor in a new PR.